File: backends/Contracts/GeoRefDataService.Default/Handlers/GetFromMem.cs
// -----------------------------------------------------------------------
// <copyright file="GetFromMem.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Contracts.GeoRefDataService.Default.Handlers;

// ReSharper disable AccessToStaticMemberViaDerivedType
using D2.Contracts.Handler;
using D2.Contracts.Interfaces.MemoryCacheService;
using D2.Contracts.Result;
using D2.Services.Protos.Geo.V1;
using H = D2.Contracts.Interfaces.GeoRefDataService.IGeoRefDataService.IGetFromMemHandler;
using I = D2.Contracts.Interfaces.GeoRefDataService.IGeoRefDataService.GetFromMemInput;
using O = D2.Contracts.Interfaces.GeoRefDataService.IGeoRefDataService.GetFromMemOutput;

/// <summary>
/// Handler for getting georeference data from memory cache.
/// </summary>
public class GetFromMem : BaseHandler<GetFromMem, I, O>, H
{
    private readonly IMemoryCacheService.IGetHandler<GetReferenceDataResponse> r_memoryCacheGet;

    /// <summary>
    /// Initializes a new instance of the <see cref="GetFromMem"/> class.
    /// </summary>
    ///
    /// <param name="memoryCacheGet">
    /// The memory cache get handler.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public GetFromMem(
        IMemoryCacheService.IGetHandler<GetReferenceDataResponse> memoryCacheGet,
        IHandlerContext context)
        : base(context)
    {
        r_memoryCacheGet = memoryCacheGet;
    }

    /// <summary>
    /// Executes the handler to get georeference data from memory cache.
    /// </summary>
    ///
    /// <param name="input">
    /// The input parameters for the handler.
    /// </param>
    /// <param name="ct">
    /// The cancellation token.
    /// </param>
    ///
    /// <returns>
    /// The result of the get operation, containing the georeference data if found.
    /// </returns>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        const string cache_key = "GeoRefData";

        var getR = await r_memoryCacheGet.HandleAsync(new(cache_key), ct);

        if (getR.CheckSuccess(out var output))
        {
            return D2Result<O?>.Ok(new(output!.Value!), traceId: TraceId);
        }

        return D2Result<O?>.NotFound(traceId: TraceId);
    }
}

File: backends/node/services/geo/geo-client/src/handlers/q/get-from-mem.ts
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import type { GeoRefData } from "@d2/protos";
import type { MemoryCacheStore } from "@d2/cache-memory";
import { GEO_CACHE_KEYS } from "../../cache-keys.js";
import { Queries } from "../../interfaces/index.js";

type Input = Queries.GetFromMemInput;
type Output = Queries.GetFromMemOutput;

/**
 * Handler for getting georeference data from the in-memory cache.
 * Mirrors D2.Geo.Client.CQRS.Handlers.Q.GetFromMem in .NET.
 */
export class GetFromMem extends BaseHandler<Input, Output> implements Queries.IGetFromMemHandler {
  override get redaction() {
    return Queries.GET_FROM_MEM_REDACTION;
  }

  private readonly store: MemoryCacheStore;

  constructor(store: MemoryCacheStore, context: IHandlerContext) {
    super(context);
    this.store = store;
  }

  protected async executeAsync(_input: Input): Promise<D2Result<Output | undefined>> {
    const value = this.store.get<GeoRefData>(GEO_CACHE_KEYS.REFDATA);
    if (value === undefined) {
      return D2Result.notFound();
    }
    return D2Result.ok({ data: { data: value } });
  }
}

export type { GetFromMemInput, GetFromMemOutput } from "../../interfaces/q/get-from-mem.js";
